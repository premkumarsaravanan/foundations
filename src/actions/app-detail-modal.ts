import { actionCreator } from '../utils/actions'
import ActionTypes from '../constants/action-types'

export const setAppDetailModalStateView = actionCreator<void>(ActionTypes.SET_APP_DETAIL_MODAL_STATE_VIEW)
export const setAppDetailModalStatePermission = actionCreator<void>(ActionTypes.SET_APP_DETAIL_MODAL_STATE_PERMISSION)
export const setAppDetailModalStateViewConfirm = actionCreator<void>(ActionTypes.SET_APP_DETAIL_MODAL_STATE_CONFIRM)
export const setAppDetailModalStateInstallSuccess = actionCreator<void>(ActionTypes.SET_APP_DETAIL_MODAL_STATE_SUCCESS)
