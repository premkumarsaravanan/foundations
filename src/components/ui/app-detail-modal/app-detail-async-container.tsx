import { connect } from 'react-redux'
import { ReduxState } from '@/types/core'
import AsyncContainer from '@/components/ui/async-container'

export const mapStateToProps = (state: ReduxState) => {
  const { loading, error, appDetailData } = state.appDetail

  return {
    loading,
    error,
    data: appDetailData
  }
}

const DetailAsyncContainer = connect(mapStateToProps)(AsyncContainer)

export default DetailAsyncContainer
