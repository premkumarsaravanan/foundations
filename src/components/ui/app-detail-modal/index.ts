export { default } from './app-detail-modal'
