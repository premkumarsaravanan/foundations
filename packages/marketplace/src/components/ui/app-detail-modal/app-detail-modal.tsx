import * as React from 'react'
import { Modal, ModalProps } from '@reapit/elements'
import { connect } from 'react-redux'
import { setAppDetailModalStateBrowse } from '@/actions/app-detail-modal'
import AppDetailInner from './app-detail-inner'
import AppDetailAsyncContainer from './app-detail-async-container'

export interface ActionDetailModalMappedAction {
  setAppDetailModalStateBrowse: () => void
}

export type AppDetailModalProps = Pick<ModalProps, 'visible' | 'afterClose'> & ActionDetailModalMappedAction

const mapDispatchToProps = (dispatch: any): ActionDetailModalMappedAction => ({
  setAppDetailModalStateBrowse: () => dispatch(setAppDetailModalStateBrowse()),
})

export const handleAfterClose = (setAppDetailModalStateBrowse: () => void, afterClose?: () => void) => () => {
  if (afterClose) {
    afterClose()
  }
  setAppDetailModalStateBrowse()
}

export const AppDetailModal: React.FunctionComponent<AppDetailModalProps> = ({
  visible = true,
  afterClose,
  setAppDetailModalStateBrowse,
}) => {
  return (
    <Modal visible={visible} afterClose={handleAfterClose(setAppDetailModalStateBrowse, afterClose)} renderChildren>
      <AppDetailAsyncContainer>
        <AppDetailInner afterClose={afterClose} />
      </AppDetailAsyncContainer>
    </Modal>
  )
}

const AppDetailModalWithConnect = connect(null, mapDispatchToProps)(AppDetailModal)

export default AppDetailModalWithConnect
