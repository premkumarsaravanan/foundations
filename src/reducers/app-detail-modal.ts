import { Action } from '../types/core'
import { isType } from '../utils/actions'
import {
  setAppDetailModalStateView,
  setAppDetailModalStatePermission,
  setAppDetailModalStateViewConfirm
} from '../actions/app-detail-modal'

export type AppDetailModalState = 'VIEW_DETAIL' | 'VIEW_PERMISSION' | 'VIEW_CONFIRM_INSTALL'

export const defaultState: AppDetailModalState = 'VIEW_DETAIL'

const appDetailModalReducer = (state: AppDetailModalState = defaultState, action: Action<any>): AppDetailModalState => {
  if (isType(action, setAppDetailModalStateView)) {
    return 'VIEW_DETAIL'
  }

  if (isType(action, setAppDetailModalStatePermission)) {
    return 'VIEW_PERMISSION'
  }

  if (isType(action, setAppDetailModalStateViewConfirm)) {
    return 'VIEW_CONFIRM_INSTALL'
  }

  return state
}

export default appDetailModalReducer
