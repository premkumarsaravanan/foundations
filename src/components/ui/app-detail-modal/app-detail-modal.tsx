import * as React from 'react'
import { Modal, ModalProps } from '@reapit/elements'
import { connect } from 'react-redux'
import { setAppDetailModalStateView } from '@/actions/app-detail-modal'
import { appInstallDone } from '@/actions/app-install'
import AppDetailInner from './app-detail-inner'
import AppDetailAsyncContainer from './app-detail-async-container'
import { FormState, ReduxState } from '@/types/core'

export interface ActionDetailModalMappedState {
  appInstallFormState: FormState
}

export interface ActionDetailModalMappedAction {
  setAppDetailModalStateView: () => void
  appInstallDone: () => void
}

export type AppDetailModalProps = Pick<ModalProps, 'visible' | 'afterClose'> & ActionDetailModalMappedAction

const mapDispatchToProps = (dispatch: any): ActionDetailModalMappedAction => ({
  setAppDetailModalStateView: () => dispatch(setAppDetailModalStateView()),
  appInstallDone: () => dispatch(appInstallDone())
})

const mapStateToProps = (state: ReduxState): ActionDetailModalMappedState => ({
  appInstallFormState: state.appInstall.formState
})

export const handleAfterClose = (
  setAppDetailModalStateView: () => void,
  appInstallDone: () => void,
  afterClose?: () => void
) => () => {
  if (afterClose) {
    afterClose()
    appInstallDone()
  }
  setAppDetailModalStateView()
}

export const AppDetailModal: React.FunctionComponent<AppDetailModalProps> = ({
  visible = true,
  afterClose,
  setAppDetailModalStateView,
  appInstallDone
}) => {
  return (
    <Modal visible={visible} afterClose={handleAfterClose(setAppDetailModalStateView, appInstallDone, afterClose)}>
      <AppDetailAsyncContainer>
        <AppDetailInner afterClose={afterClose} />
      </AppDetailAsyncContainer>
    </Modal>
  )
}

const AppDetailModalWithConnect = connect(
  mapStateToProps,
  mapDispatchToProps
)(AppDetailModal)

export default AppDetailModalWithConnect
