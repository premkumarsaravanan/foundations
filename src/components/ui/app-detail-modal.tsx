import * as React from 'react'
import bulma from '@/styles/vendor/bulma'
import Modal, { ModalProps } from '@/components/ui/modal'
import { connect } from 'react-redux'
import { ReduxState } from '@/types/core'
import { AppDetailState } from '@/reducers/app-detail'
import Loader from '@/components/ui/loader'
import Slider, { Settings } from 'react-slick'
import { appDetailRequestData } from '@/actions/app-detail'
import carouselStyles from '../../styles/elements/carousel.scss?mod'
import ChevronLeftIcon from '@/components/svg/chevron-left'
import Alert from '@/components/ui/alert'
import '@/styles/vendor/slick.scss'

export interface AppDetailInnerProps {
  appDetailState: AppDetailState
}

export interface AppDetailModalMappedProps {
  appDetailState: AppDetailState
}

export interface AppDetailModalMappedActions {
  fetchAppDetail: (id: string) => void
}

export type AppDetailModalProps = Pick<ModalProps, 'visible' | 'afterClose'> & {}

const SlickButtonNav = ({ currentSlide, slideCount, children, ...props }) => <button {...props}>{children}</button>

export const AppDetailInner: React.FunctionComponent<AppDetailInnerProps> = ({ appDetailState }) => {
  if (appDetailState.loading) {
    return <Loader />
  }

  if (appDetailState.error) {
    return <Alert type="danger" message="Failed to fetch. Please try later." />
  }

  if (!appDetailState.appDetailData) {
    return null
  }

  const {
    media = [],
    description,
    name,
    summary,
    developer,
    homePage,
    supportEmail
  } = appDetailState.appDetailData.data

  const icon = media.filter(({ type }) => type === 'icon')[0]
  const carouselImages = media.filter(({ type }) => type === 'image')

  const settings: Settings = {
    dots: false,
    infinite: true,
    speed: 500,
    useCSS: true,
    variableWidth: false,
    prevArrow: (
      // @ts-ignore
      <SlickButtonNav>
        <ChevronLeftIcon />
      </SlickButtonNav>
    ),
    nextArrow: (
      // @ts-ignore
      <SlickButtonNav>
        <ChevronLeftIcon />
      </SlickButtonNav>
    ),
    slidesToShow: 3,
    slidesToScroll: 1
  }

  return (
    <>
      <div className={bulma.media}>
        <div className={bulma.mediaLeft}>
          <figure className={bulma.image + ' ' + bulma.is128x128}>{icon && <img src={icon.uri} />}</figure>
        </div>
        <div className={bulma.mediaContent}>
          <div className={bulma.content}>
            <h3 className={bulma.isH3} style={{ marginBottom: 0 }}>
              {name}
            </h3>
            <p>
              <small className={bulma.hasTextLink}>{developer}</small>
            </p>
            <p>{summary}</p>
          </div>
        </div>
      </div>
      {carouselImages.length > 0 && (
        <div className={carouselStyles.container}>
          <Slider {...settings}>
            {carouselImages.map(({ uri }, index) => (
              <div key={index} className={carouselStyles.slide}>
                <img src={uri} />
              </div>
            ))}
          </Slider>
        </div>
      )}
      <br />
      <p>{description}</p>
    </>
  )
}

const mapStateToProps = (state: ReduxState): AppDetailModalMappedProps => ({
  appDetailState: state.appDetail
})

const mapDispatchToProps = (dispatch: any): AppDetailModalMappedActions => ({
  fetchAppDetail: () => dispatch(appDetailRequestData('xxxx'))
})

const AppDetailInnerWithConnect = connect(
  mapStateToProps,
  mapDispatchToProps
)(AppDetailInner)

export const AppDetailModal: React.FunctionComponent<AppDetailModalProps> = ({ visible = true, afterClose }) => {
  return (
    <Modal visible={visible} afterClose={afterClose}>
      <AppDetailInnerWithConnect />
    </Modal>
  )
}

export default AppDetailModal
