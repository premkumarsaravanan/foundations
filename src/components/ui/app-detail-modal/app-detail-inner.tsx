import * as React from 'react'
import { connect } from 'react-redux'
import { AppDetailModalState } from '@/reducers/app-detail-modal'
import { ReduxState } from '@/types/core'
import AppDetail from '@/components/ui/app-detail'
import { AppDetailState } from '@/reducers/app-detail'
import AppPermission from '@/components/ui/app-permission'
import AppInstallConfirm from '@/components/ui/app-confirm-install'

export interface AppDetailInnerMappedProps {
  appDetailModalState: AppDetailModalState
  appDetailState: AppDetailState
}

export const mapStateToProps = (state: ReduxState): AppDetailInnerMappedProps => ({
  appDetailModalState: state.appDetailModal,
  appDetailState: state.appDetail
})

type AppDetailInnerProps = AppDetailInnerMappedProps & {
  afterClose?: () => void
}

export const AppDetailInner: React.FunctionComponent<AppDetailInnerProps> = ({
  appDetailModalState,
  appDetailState,
  afterClose
}) => {
  if (appDetailModalState === 'VIEW_DETAIL') {
    if (!appDetailState.appDetailData || !appDetailState.appDetailData.data) {
      return null
    }
    return <AppDetail data={appDetailState.appDetailData.data} />
  }

  if (appDetailModalState === 'VIEW_PERMISSION') {
    return <AppPermission afterClose={afterClose} />
  }

  if (appDetailModalState === 'VIEW_CONFIRM_INSTALL') {
    return <AppInstallConfirm afterClose={afterClose} />
  }

  return null
}

const AppDetailInnerWithConnect = connect(
  mapStateToProps,
  null
)(AppDetailInner)

export default AppDetailInnerWithConnect
