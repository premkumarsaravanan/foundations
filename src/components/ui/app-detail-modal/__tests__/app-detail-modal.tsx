import * as React from 'react'
import { shallow } from 'enzyme'
import toJson from 'enzyme-to-json'
import { AppDetailModal, AppDetailModalProps, handleAfterClose } from '../app-detail-modal'

const props: AppDetailModalProps = {
  visible: true,
  afterClose: jest.fn(),
  setAppDetailModalStateView: jest.fn()
}

describe('AppDetailModel', () => {
  it('should match a snapshot', () => {
    expect(toJson(shallow(<AppDetailModal {...props} />))).toMatchSnapshot()
  })
  describe('handleAfterClose have afterClose', () => {
    const afterClose = jest.fn()
    const setAppDetailModalStateView = jest.fn()
    const fn = handleAfterClose(setAppDetailModalStateView, afterClose)
    fn()
    expect(afterClose).toBeCalled()
    expect(setAppDetailModalStateView).toBeCalled()
  })
  describe('handleAfterClose doesnt have afterClose', () => {
    const afterClose = undefined
    const setAppDetailModalStateView = jest.fn()
    const fn = handleAfterClose(setAppDetailModalStateView, afterClose)
    fn()
    expect(setAppDetailModalStateView).toBeCalled()
  })
})
