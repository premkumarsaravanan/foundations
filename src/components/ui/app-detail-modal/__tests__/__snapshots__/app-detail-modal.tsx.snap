// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppDetailModel should match a snapshot 1`] = `
<Modal
  afterClose={[Function]}
  visible={true}
>
  <Connect(AsyncContainer)>
    <Connect(Component)
      afterClose={[MockFunction]}
    />
  </Connect(AsyncContainer)>
</Modal>
`;
